// @appbaseio/reactivesearch/lib/components/basic/StateProvider.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _propTypes = require('prop-types');

var _helper = require('@appbaseio/reactivecore/lib/utils/helper');

var _utils = require('../../utils');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var filterProps = function filterProps(props) {
	return _extends({}, props, {
		props: props.componentProps
	});
};

var filterByComponentIds = function filterByComponentIds(state) {
	var props = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
	var componentIds = props.componentIds;

	if (typeof componentIds === 'string') {
		var _ref;

		return _ref = {}, _ref[componentIds] = state[componentIds], _ref;
	}
	if (componentIds instanceof Array) {
		var filteredState = {};
		componentIds.forEach(function (componentId) {
			filteredState[componentId] = state[componentId];
		});
		return filteredState;
	}
	return state;
};

var StateProvider = function (_Component) {
	_inherits(StateProvider, _Component);

	function StateProvider(props) {
		_classCallCheck(this, StateProvider);

		var _this = _possibleConstructorReturn(this, _Component.call(this, props));

		_this.state = {
			searchState: (0, _helper.getSearchState)(filterProps(props))
		};
		return _this;
	}

	StateProvider.getDerivedStateFromProps = function getDerivedStateFromProps(props) {
		return {
			searchState: (0, _helper.getSearchState)(filterProps(props))
		};
	};

	StateProvider.prototype.shouldComponentUpdate = function shouldComponentUpdate(nextProps, nextState) {
		// Only apply when componentIds is defined
		if (!nextProps.strict || JSON.stringify(nextState) !== JSON.stringify(this.state)) {
			return true;
		}
		return false;
	};

	StateProvider.prototype.render = function render() {
		var searchState = this.state.searchState;

		return (0, _utils.getComponent)({ searchState: searchState }, this.props);
	};

	return StateProvider;
}(_react.Component);

StateProvider.defaultProps = {
	strict: true
};
StateProvider.propTypes = {
	componentIds: (0, _propTypes.oneOfType)([_propTypes.string, (0, _propTypes.arrayOf)(_propTypes.string)]),
	strict: _propTypes.bool
};

var mapStateToProps = function mapStateToProps(state, props) {
	return {
		selectedValues: filterByComponentIds(state.selectedValues, props),
		queryLog: filterByComponentIds(state.queryLog, props),
		dependencyTree: filterByComponentIds(state.dependencyTree, props),
		componentProps: filterByComponentIds(state.props, props),
		hits: filterByComponentIds(state.hits, props),
		aggregations: filterByComponentIds(state.aggregations, props),
		isLoading: filterByComponentIds(state.isLoading, props),
		error: filterByComponentIds(state.error, props)
	};
};

exports.default = (0, _utils.connect)(mapStateToProps, null)(StateProvider);
